crates/llms/src/anthropic/types_stream.rs

```diff
@@ -0,0 +1,395 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "AnthropicStreamError: {:?}", self.error)
+    }
+}
+
+impl From<reqwest_eventsource::Error> for AnthropicStreamError {
+    fn from(e: reqwest_eventsource::Error) -> Self {
+        AnthropicStreamError {
+            event_type: "error".to_string(),
+            error: ErrorPayload {
+                error_type: "reqwest_eventsource_error".to_string(),
+                message: e.to_string(),
+            },
+        }
+    }
+}
+
+impl From<serde_json::Error> for AnthropicStreamError {
+    fn from(e: serde_json::Error) -> Self {
+        AnthropicStreamError {
+            event_type: "error".to_string(),
+            error: ErrorPayload {
+                error_type: "serde_json_error".to_string(),
+                message: e.to_string(),
+            },
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct ErrorPayload {
+    #[serde(rename = "type")]
+    error_type: String,
+    message: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct MessageDelta {
+    pub stop_reason: Option<StopReason>,
+    pub stop_sequence: Option<String>,
+}
+
+/// Convert the stream of Anthropic [`MessageCreateStreamResponse`] into a stream of `OpenAI` compatible [`async_openai::types::CreateChatCompletionStreamResponse`].
+///
+/// Except for differences in the stream packet formats, the core difference are:
+///
+///  +---------------------------------------------------------+---------------------------------------------------------+
+///  | Anthropic                                               | `OpenAI`                                                  |
+///  +---------------------------------------------------------+---------------------------------------------------------+
+///  | Only first packet for a specific tool has tool metadata | All packets for a tool have tool metadata               |
+///  |                                                         |                                                         |
+///  | Initial message has initial usage details. Last message | Last message has usage details.                         |
+///  | has additional usage details.                           |                                                         |
+///  |                                                         |                                                         |
+///  | Tool packets have no out of order protection            | Provides numbering for out of order tool packets        |
+///  +---------------------------------------------------------+---------------------------------------------------------+
+///
+#[allow(clippy::too_many_lines)]
+pub fn transform_stream(
+    mut stream: Pin<
+        Box<dyn Stream<Item = Result<MessageCreateStreamResponse, AnthropicStreamError>> + Send>,
+    >,
+    model: String,
+) -> ChatCompletionResponseStream {
+    let transformed_stream = stream! {
```

Comment:
The stream! macro breaks most code completion within the macro block, so I'd prefer its only used for small code blocks. Could this be rewritten using native Rust instead? Perhaps using `stream::unfold`?